### PR TITLE
Handle yaml merge keys correcly.

### LINKED ISSUE #886: YAML anchor key merge
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

pip install esphome
arch linux

**ESP (ESP32/ESP8266, Board/Sonoff):**

shelly 2.5 
```yaml
esphome:
  platform: ESP8266
  board: esp01_1m
```

**Affected component:**

**Description of problem:**

YAML anchor tags doesn't work as expected.

`esphome shelly_25.yaml config`

Output:
> ERROR Error while reading config: Invalid YAML syntax. Please see YAML syntax reference or use an online YAML syntax validator:
> 
> duplicate key: "name"

Expected result: the latter should override the former.

**Problem-relevant YAML-configuration entries:**
```yaml

binary_sensor:
  - &button1
    platform: gpio
    name: "button 1"
    pin:
      number: GPIO2

switch:
  <<: *button1
  name: "override"
```

**Logs (if applicable):**

```
PASTE DEBUG LOG HERE

```

**Additional information and things you've tried:**

### PR DESCRIPTION
**Pull request by edge90**
## Description:

**Related issue (if applicable):** fixes https://github.com/esphome/issues/issues/886

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/yaml_util.py [hunk 1]
```diff
@@ -130,50 +130,45 @@ def construct_yaml_seq(self, node):
...
                 index += 1
             else:
                 index += 1
-        if pre_merge:
-            node.value = pre_merge + node.value
-        if post_merge:
-            node.value = node.value + post_merge
+        if merge:
+            def determine(key, value, haystack):
+                return key.value in haystack
```
Comment (OttoWinter) [inline esphome/yaml_util.py]: This line doesn't look right - it will access the attribute .value in `key` (as in the parameter called `value` literally). The parameter `value` to this function is unused.

File: tests/merge_anchors.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+substitutions: &esphome
```
Comment (OttoWinter) [inline tests/merge_anchors.yaml]: Unfortunately, this file needs to be removed - the problem with the tests folder is that each test file adds a significant amount of time to the CI checks and thus needs to be limited.

You can try to add YAML anchors to the existing test files though.

**Review by OttoWinter — COMMENTED** [review]
Please also explain how you came up with this code change - it's hard to analyze this change (and check it for correctness) from the diff alone.

Is this based on the updated PyYAML codebase? How does it handle duplicate keys in *normal* dicts (without merge keys)? The new code should definitely still raise an error if there are normal duplicate keys.

File: esphome/yaml_util.py (hunk 1 above)
Comment (edge90) [reply esphome/yaml_util.py]: Without having the code infront of me I'm pretty sure that key.value contains the the key name, like platform, pin etc. value.value would contain gpio, GPIO0 etc.

The whole point of that is to not add keys that should be overridden in the next step.

File: tests/merge_anchors.yaml (hunk 2 above)
Comment (edge90) [reply tests/merge_anchors.yaml]: Ok, will do!

**Comment by edge90** [comment]
I was trying to avoid config/code duplication and wanted to solve it by using anchors. The normal python implemention does not complain if you have duplicate keys. This implemention still does if you omit the merge "operator".

And yes, it's based on the PyYAML implemention which I tweaked to handle the merge stuff (as PyYAML doesn't need to do as it doesn't care about duplicate keys)

File: tests/merge_anchors.yaml (hunk 2 above)
Comment (edge90) [reply tests/merge_anchors.yaml]: Removed, misspelled your name though :/

**Comment by OttoWinter** [comment]
Hi, thanks a lot!

I looked over the implementation and the [YAML spec](https://yaml.org/type/merge.html) and found one edge case:

If a merge sequence is used, the YAML spec says the merge entries defined earlier in the sequence should override those later in the sequence. The current code returns a duplicate-key error for this input:
```yaml
.test: &test
  initial_value: 'value1'
.test2: &test2
  initial_value: 'value2'

globals:
  - id: something
    type: 'int'
    <<: [*test, *test2]
```

I tweaked the merge code a bit to fix that edge case (probably not too many people know about that merge sequence feature anyway, but still) and now it is fixed.

I've pushed to code, if you're ok with the changes I'd be happy to merge.

**Comment by edge90** [comment]
Fantastic, good catch! 

Looks fine to me so you can go ahead and merge :)
